microservice/src/test/java/de/gsi/microservice/concepts/RoundTripAndNotifyTests.java

```diff
@@ -0,0 +1,433 @@
+ * (ie. JIT, whether services are identified by single characters etc.)
+ */
+class RoundTripAndNotifyTests {
+    // private static final String SUB_TOPIC = "x";
+    private static final String SUB_TOPIC = "<domain>/<property>?<filter>#<ctx> - a very long topic to test the dependence of pub/sub pairs on topic lengths";
+    private static final byte[] SUB_DATA = "D".getBytes(StandardCharsets.UTF_8); // custom minimal data
+    private static final byte[] CLIENT_ID = "C".getBytes(StandardCharsets.UTF_8); // client name
+    private static final byte[] WORKER_ID = "W".getBytes(StandardCharsets.UTF_8); // worker-service name
+    private static final byte[] PUBLISH_ID = "P".getBytes(StandardCharsets.UTF_8); // publish-service name
+    private static final byte[] SUBSCRIBER_ID = "S".getBytes(StandardCharsets.UTF_8); // subscriber name
+    private static int SAMPLE_SIZE = 10_000;
+    private static int SAMPLE_SIZE_PUB = 100_000;
+
+    public static void main(String[] args) {
+        if (args.length == 1) {
+            SAMPLE_SIZE = Integer.parseInt(args[0]);
+            SAMPLE_SIZE_PUB = 10 * SAMPLE_SIZE;
+        }
+
+        Thread brokerThread = new Thread(new Broker());
+        Thread workerThread = new Thread(new Worker());
+        Thread publishThread = new Thread(new PublishWorker());
+        Thread pubDealerThread = new Thread(new PublishDealerWorker());
+        Thread directDealerThread = new Thread(new PublishDirectDealerWorker());
+
+        brokerThread.start();
+        workerThread.start();
+        publishThread.start();
+        pubDealerThread.start();
+        directDealerThread.start();
+
+        Thread clientThread = new Thread(new Client());
+        clientThread.start();
+
+        try {
+            clientThread.join();
+            workerThread.interrupt();
+            brokerThread.interrupt();
+            publishThread.interrupt();
+            pubDealerThread.interrupt();
+            directDealerThread.interrupt();
+
+            // wait for threads to finish
+            workerThread.join();
+            publishThread.join();
+            pubDealerThread.join();
+            directDealerThread.join();
+            brokerThread.join();
+        } catch (InterruptedException e) {
+            // finishes tests
+        }
+        System.out.println("finished tests");
+    }
+
+    private static void measure(final String topic, final int nExec, final Runnable... runnable) {
+        final long start = System.currentTimeMillis();
+
+        for (Runnable run : runnable) {
+            for (int i = 0; i < nExec; i++) {
+                run.run();
+            }
+        }
+
+        final long stop = System.currentTimeMillis();
+        System.out.printf("%-40s:  %10d calls/second\n", topic, (1000 * nExec) / (stop - start));
+    }
+
+    static class Broker implements Runnable {
+        private static final int TIMEOUT = 1000;
+        @Override
+        public void run() {
```

Comment:
Codacy found an issue: [The method 'run()' has an NPath complexity of 2606, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6307742)